orchagent/aclorch.cpp

```diff
@@ -2142,6 +2221,100 @@ void AclRuleMirror::onUpdate(SubjectType type, void *cntx)
     }
 }
 
+AclRuleUnderlaySetDhcp::AclRuleUnderlaySetDhcp(AclOrch *aclOrch, string rule, string table, MetaDataMgr* m_metaDataMgr, bool createCounter):
+    AclRule(aclOrch, rule, table, createCounter),
+    table_id(table),
+    m_metaDataMgr(m_metaDataMgr)
+{
+}
+
+uint32_t AclRuleUnderlaySetDhcp::getDscpValue() const
+{
+    return cachedDscpValue;
+}
+
+uint32_t AclRuleUnderlaySetDhcp::getMetadata() const
+{
+    return cachedMetadata;
+}
+
+bool AclRuleUnderlaySetDhcp::validateAddAction(string attr_name, string _attr_value)
+{
+    SWSS_LOG_ENTER();
+
+    string attr_value = to_upper(_attr_value);
+
+    sai_object_id_t table_oid = m_pAclOrch->getTableById(table_id);
+    auto aclTable = m_pAclOrch->getTableByOid(table_oid);
+    string type = aclTable->type.getName();
+    string key = table_id + ":" + m_id;
+    // we handle the allocation of metadata for here. based on SET_DSCP action, we check if a metadata is already allocated then we reuse it
+    // otherwise we allocate a new metadata. This metadata is then set an the action for the Rule of this table. We also cache the SET_DSCP
+    // value and the allocated metadata in a the rule structure itself so that when we go to addRule we can use these to add the
+    // egr_set_dscp rule
+    if (attr_name == ACTION_DSCP && (type == TABLE_TYPE_MARK_META || type == TABLE_TYPE_MARK_META_V6))
```

Comment:
What ports are the Egress table bound to? Is it always the same with the Ingress table?

Reply:
The Egress table is bound to a set of ports which contains all the ports associated with tables of type UNDERLAY_SET_DSCP and UNDERLAY_SET_DSCPV6. There is a test **test_OverlayBothv4v6TableCreationDeletion** and **test_OverlayBothv4v6TableSameintfCreationDeletion** test these scenarios.
e.g. if one table has ports [eth0,eth1,eth2] and other has [eth2,eth3,eth4]
The egress table would be bound to [eth0,eth1,eth2,eth3,eth4].
If one of the table is removed only the ports associated with the table are removed except those which are used by other others. In the above mentioned case  removal of first table would remove only eth0 and eth1. eth2 would be retained because second table is still bound to it.

Reply:
The port binding may not be good enough for the prod scenario.
For example, the packet that we are interested is ingressed from `eth0` and `eth1` and egressed from `eth10`. So the ingress table should be bound to `eth0` and `eth1`, and egress table should be bound to `eth10` only. In current design, the egress table is always bound to `eth1`, `eth2` and `eth10`. The concern is the DSCP of unexpected packet is changed.

Reply:
In order for the Egr_Set_dscp to work similar to everflow, it needs to be bound to all upstream or downsteam ports. There are 2 ways to do that. 
1) by default statically set EGR_SET_DCP table to be bound to all ports in the Orchagent code.
2) While configuring, Associate the UNDERLAY_SET_DSCP/v6 with both upstream and downstream ports as it is done in prod scenarios with everflow.
While option 1 is more idiot-proof, it requires midofication in portMgr to expose list of active/Inactive ports along with their type.( since we dont need this on Mgmt or cpu ports, the filtering on the port type would be required.) This makes the whole implementation hackey. Another downside is we are  consuming table membership even when one may not need to and the user of this table has no choice to select the port association for this table in the future.
I spent some time exploring the first option and concluded that while its is possible, it make the whole solution very specific to one use case.
The dowside of option 2 is  that user has to account for and be aware the fact that the traffic may egress on ports not asociated with  UNDERLAY_SET_DSCP/v6. But for the prod scenario, we would configure it with both upstream and downstream ports just as we do with everflow.